modules/flann/include/opencv2/flann/kmeans_index.h

```diff
@@ -297,24 +295,15 @@ class KMeansIndex : public NNIndex<Distance>
 
             for( int i = begin; i<end; ++i)
             {
-                DistanceType sq_dist = distance(dataset[indices[i]], dcenters[0], veclen);
-                int new_centroid = 0;
+                sq_dists[i] = distance(dataset[indices[i]], dcenters[0], veclen);
+                new_centroids[i] = 0;
                 for (int j=1; j<branching; ++j) {
                     DistanceType new_sq_dist = distance(dataset[indices[i]], dcenters[j], veclen);
-                    if (sq_dist>new_sq_dist) {
-                        new_centroid = j;
-                        sq_dist = new_sq_dist;
+                    if (sq_dists[i]>new_sq_dist) {
+                        new_centroids[i] = j;
+                        sq_dists[i] = new_sq_dist;
```

Comment:
It is better to make calculations using local variables and save them in global memory in the end.

```
for( int i = begin; i<end; ++i)
{
    DistanceType sq_dist = ...
    int new_centroid = 0;
    ...
    sq_dists[i] = sq_dist;
    new_centroids[i] = new_centroid;
}
```

Reply:
Honestly I thought compilers could do this optimization automatically. I guess not...